internal/campaigns/run_steps.go

```diff
@@ -297,3 +308,57 @@ func probeImageForShell(ctx context.Context, image string) (shell, tempfile stri
 	// admit defeat and return. At least err is already in place.
 	return
 }
+
+type StepFailedErr struct {
+	Run       string
+	Container string
+
+	TmpFilename string
+
+	Args   []string
+	Stdout string
+	Stderr string
+
+	Err error
+}
+
+func (e StepFailedErr) Error() string {
+	var out strings.Builder
+
+	fmtRun := func(run string) string {
+		lines := strings.Split(run, "\n")
+		if len(lines) == 1 {
+			return lines[0]
+		}
+		return lines[0] + fmt.Sprintf("\n\t(... and %d more lines)", len(lines)-1)
+	}
+
+	out.WriteString(fmt.Sprintf("run: %s\ncontainer: %s\n", fmtRun(e.Run), e.Container))
+
+	printOutput := func(output string) {
+		for _, line := range strings.Split(output, "\n") {
+			if e.TmpFilename != "" {
+				line = strings.ReplaceAll(line, e.TmpFilename+": ", "")
+			}
+			out.WriteString("\t" + line + "\n")
+		}
+	}
+
+	if len(e.Stdout) > 0 {
+		out.WriteString("\nstandard out:\n")
+		printOutput(e.Stdout)
+	}
+
+	if len(e.Stderr) > 0 {
+		out.WriteString("\nstandard error:\n")
+		printOutput(e.Stderr)
+	}
+
+	if exitErr, ok := e.Err.(*exec.ExitError); ok {
+		out.WriteString(fmt.Sprintf("\nCommand failed with exit code %d.", exitErr.ExitCode()))
```

Comment:
Seems like we always set the exit code to 1? If so, I think it would be confusing

Reply:
Well, we always set it to 1 in _our `src campaigns [apply|preview]`_, right? But take a look at the screenshot in the PR description: the step fails with exit code 127.

Reply:
Ah, I see, cool